clang/test/Analysis/fread.c

```diff
@@ -0,0 +1,443 @@
+      // FIXME: The store should have returned a partial binding for the 17th byte of the buffer, which is the 2nd byte of the previous int.
+      // This byte should have been initialized by the 'fread' earlier. However, the Store lies to us and says it's uninitialized.
+      clang_analyzer_dump(c[4+13]); // expected-warning{{1st function call argument is an uninitialized value}} should be initialized.
+      clang_analyzer_dump(c[4+16]); // This should be the first byte that 'fread' leaves uninitialized. This should raise the uninit read diag.
+    } else {
+      clang_analyzer_dump(buffer[0]); // expected-warning{{1 S32b}} ok
+      clang_analyzer_dump(buffer[1]); // expected-warning{{conj_}} ok
+      clang_analyzer_dump(buffer[2]); // expected-warning{{conj_}} ok
+      clang_analyzer_dump(buffer[3]); // expected-warning{{conj_}} ok
+      clang_analyzer_dump(buffer[4]); // expected-warning{{conj_}} ok, but an uninit warning would be also fine.
+      clang_analyzer_dump(buffer[5]); // expected-warning{{1st function call argument is an uninitialized value}} ok
+    }
+    fclose(fp);
+  }
+}
+
+void test_whole_elements_read(void) {
+  clang_analyzer_dump(sizeof(int)); // expected-warning {{4 S32b}}
+
+  int buffer[100];
+  buffer[0] = 1;
+  buffer[15] = 2;
+  buffer[16] = 3;
+  FILE *fp = fopen("/home/test", "rb+");
+  if (fp) {
+    // 3*20: 60 bytes read; which is basically 15 integers.
+    if (20 == fread(buffer + 1, 3, 20, fp)) {
+      clang_analyzer_dump(buffer[0]);  // expected-warning{{1 S32b}}
+      clang_analyzer_dump(buffer[15]); // expected-warning{{conj_}}
+      clang_analyzer_dump(buffer[16]); // expected-warning{{3 S32b}}
+      clang_analyzer_dump(buffer[17]); // expected-warning{{1st function call argument is an uninitialized value}}
+    } else {
+      clang_analyzer_dump(buffer[0]);  // expected-warning{{1 S32b}}
+      clang_analyzer_dump(buffer[15]); // expected-warning{{conj_}}
+      clang_analyzer_dump(buffer[16]); // expected-warning{{3 S32b}}
+      clang_analyzer_dump(buffer[17]); // expected-warning{{1st function call argument is an uninitialized value}}
+    }
+    fclose(fp);
+  }
+}
+
+void test_unaligned_start_read(void) {
+  clang_analyzer_dump(sizeof(int)); // expected-warning {{4 S32b}}
+
+  int buffer[100];
+  buffer[0] = 3;
+  buffer[1] = 4;
+  buffer[2] = 5;
+  char *asChar = (char*)buffer;
+
+  FILE *fp = fopen("/home/test", "rb+");
+  if (fp) {
+    // We have an 'int' binding at offset 0 of value 3.
+    // We read 4 bytes at byte offset: 1,2,3,4.
+    if (4 == fread(asChar + 1, 1, 4, fp)) {
+      clang_analyzer_dump(buffer[0]); // expected-warning{{3 S32b}} FIXME Reading a 8 bit 'char' should not result in a 'S32b' value.
+      clang_analyzer_dump(buffer[1]); // expected-warning{{conj_}}
+      clang_analyzer_dump(buffer[2]); // expected-warning{{5 S32b}}
+
+      clang_analyzer_dump(asChar[0]); // expected-warning{{3 S32b}} FIXME Reading a 8 bit 'char' should not result in a 'S32b' value.
```

Comment:
Does it matter that clang_analyzer_dump has argument type `int`?

Reply:
It shouldn't as the lvalue to rvalue conversion (aka. the load) operation should have been done before the call.
In contrast to this, CSA appears to contradict this assumption, and the `clang_analyzer_dump_char()` would have dumped `3 S8b`, which is the correct solution for little-endian systems - like the one I pinned as the target

I fixed this by dumping these as chars, and refined the expectation messages to meet this.